File: 0106 Construct Binary Tree from Inorder and Postorder Traversal/0106.py
class Solution:
    def build(self, inorder, postorder, l, r, mp):
        if l > r: return None

        value = postorder[self.postorder_index]
        self.postorder_index -= 1

        node = TreeNode(value)
        i = mp[value]

        node.right = self.build(inorder, postorder, i+1, r, mp)
        node.left = self.build(inorder, postorder, l, i-1, mp)

        return node


    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        n = len(inorder)
        self.postorder_index = n-1
        mp = {inorder[i] : i for i in range(n)}
        return self.build(inorder, postorder, 0, n-1, mp)

File: 0106 Construct Binary Tree from Inorder and Postorder Traversal/0106.cpp
class Solution {
public:
    TreeNode* build(unordered_map<int, int>& mp, vector<int>& postorder, int l, int r, int& postorderIndex){
        if (l > r) return nullptr;

        int value = postorder[postorderIndex--];
        TreeNode* node = new TreeNode(value);
        int i = mp[value];
        
        node->right = build(mp, postorder, i+1, r, postorderIndex);
        node->left = build(mp, postorder, l, i-1, postorderIndex);

        return node;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> mp;
        int n = inorder.size();

        for (int i = 0; i<n; i++){
            mp[inorder[i]] = i;
        }
        
        int postorderIndex = n-1;
        return build(mp, postorder, 0, n-1, postorderIndex);
    }
};
